File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
#include <string>
using namespace std;
int main()
{
int t;
cin >> t;
while(t > 0) {
t--;
string str;
cin >> str;
stack<char> s;
int check = 0;
for (int i = 0; i < str.length(); i++) {
if (str[i] == '(') {
s.push(str[i]);
}
else {
if (s.empty() || s.top() == ')') {
check = 1;
break;
}
else {
s.pop();
}
}
}
if (!s.empty() || check) {
cout << "NO" << '\n';
}
else {
cout << "YES" << '\n';
}
}
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.*;

class Main{
    public static void main(String[] args)throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));        
        int n = Integer.parseInt(br.readLine());

        StringBuilder sb = new StringBuilder();
        
        for(int i = 0; i<n; i++){
            Stack<Character> stack = new Stack<>();
            String str = br.readLine();
            boolean flag = true;
            
            for(int j = 0; j<str.length(); j++){
                char c = str.charAt(j);
                if(c=='(') stack.push(c);
                else{
                    if(stack.isEmpty()) flag = false;
                    else stack.pop();
                }
            }

            if(!flag || !stack.isEmpty()) sb.append("NO\n");
            else sb.append("YES\n");
        }

        System.out.println(sb);

    }
}